From 18d53e45103d3585e101a13be28943e39141afed Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:44 -0400
Subject: [fs] nfs: use blocking page_group_lock in add_request

Message-id: <1411050057-12417-82-git-send-email-steved@redhat.com>
Patchwork-id: 96233
O-Subject: [RHEL7.1 PATCH 81/94] nfs: use blocking page_group_lock in add_request
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Weston Andros Adamson <dros@primarydata.com>

__nfs_pageio_add_request was calling nfs_page_group_lock nonblocking, but
this can return -EAGAIN which would end up passing -EIO to the application.

There is no reason not to block in this path, so change the two calls to
do so. Also, there is no need to check the return value of
nfs_page_group_lock when nonblock=false, so remove the error handling code.

Signed-off-by: Weston Andros Adamson <dros@primarydata.com>
Reviewed-by: Peng Tao <tao.peng@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: bfd484a5606d6a0379a0a2f04251b1e5c1f8995c
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/pagelist.c b/fs/nfs/pagelist.c
index efc9c91..080c88f 100644
--- a/fs/nfs/pagelist.c
+++ b/fs/nfs/pagelist.c
@@ -857,13 +857,8 @@ static int __nfs_pageio_add_request(struct nfs_pageio_descriptor *desc,
  struct nfs_page *subreq;
  unsigned int bytes_left = 0;
  unsigned int offset, pgbase;
- int ret;
 
- ret = nfs_page_group_lock(req, true);
- if (ret < 0) {
-  desc->pg_error = ret;
-  return 0;
- }
+ nfs_page_group_lock(req, false);
 
  subreq = req;
  bytes_left = subreq->wb_bytes;
@@ -886,11 +881,7 @@ static int __nfs_pageio_add_request(struct nfs_pageio_descriptor *desc,
    if (desc->pg_recoalesce)
     return 0;
    /* retry add_request for this subreq */
-   ret = nfs_page_group_lock(req, true);
-   if (ret < 0) {
-    desc->pg_error = ret;
-    return 0;
-   }
+   nfs_page_group_lock(req, false);
    continue;
   }
 
-- 
1.7.1